### PR TITLE
Update UDS/UDI storage implementation documentation

### PR DESCRIPTION
**Pull request by mchack-work**
An attempt at updating the documentation in the tk1 and uds cores about the new storage of the UDI and UDS.

Updates the comment in the patch script to describe what it does and how to run instead of focusing on the storage solution.

If I misunderstood something there's no need to comment and ask me to update, just commit a correction. If there are things up for discussion, please comment.

### CODE REVIEW

File: hw/application_fpga/core/tk1/README.md [hunk 1]
```diff
@@ -111,9 +111,21 @@ secret for any secrets it needs to perform its intended use case.
...
+"udi\_rom\_idx".
+
+We use bit indexing from all of the 32 LUTs to store each bit of the
+word, i.e., the first UDS word consists of bit 0 from each 32 LUTs,
+the second word consists of bit 1 from each of the 32 LUTs, and so
+on... for a maximum storage of 16 words.
+
+The UDI does not occupy the entire LUT4 instance and to preserve
+hardware resources and not having to care about all the address bits
+the UDI should be initialized to be repeated over the unused portion
```
Comment (dehanj) [inline hw/application_fpga/core/tk1/README.md:126]: A bit unsure over this sentence " should be initialized to be repeated over ". Should "initialized" be removed? 
Think it is a complicated sentence.

Comment (mchack-work) [reply hw/application_fpga/core/tk1/README.md]: Fixed.

**Review by dehanj — APPROVED** [review]

File: hw/application_fpga/core/tk1/README.md [hunk 2]
```diff
@@ -111,9 +111,21 @@ secret for any secrets it needs to perform its intended use case.
...
+We use bit indexing from all of the 32 LUTs to store each bit of the
+word, i.e., the first UDI word consists of bit 0 from each 32 LUTs,
+the second word consists of bit 1 from each of the 32 LUTs, and so
+on... for a maximum storage of 16 words, but we use only 2.
+
+The UDI does not occupy the entire LUT4 instance and to preserve
+hardware resources and not having to care about all the address bits
+the UDI should be repeated over the unused portion of the LUT4
+instance. This eliminates the need to drive the three MSB pins into
+the LUT4 while still achieving the correct output.
```
Comment (secworks) [inline hw/application_fpga/core/tk1/README.md:129]: I don't agree on the preserve HW resources phrasing. In fact we are wasting 32 * 14 bits by using this setup. We could have used four LUT4 and then muxes and a couple of registers. But that would affect performance and latency.

I would rephrase to something that:
- States that 32 explictly instantiated are used, one for each bit in the data word the CPU reads.
- Each LUT4 can store 16-bits. But since the UDI is 64 bits, only two of the 16 bits are actually used
- The LUT4 are populated with a default bit pattern, but the pattern is replaced, patched in a separate stage of the generation of the FPGA bitstream. This allows us to have device unique UDIs without generating a new FPGA bitstream for each device.

File: hw/application_fpga/core/uds/README.md [hunk 3]
```diff
@@ -31,4 +31,15 @@ Any access to another address will be ignored by the core.
...
 
-The UDS words are implemented using discrete registers.
+The 8 UDS words are stored using 32 named LUT4s identified by
+"uds\_rom\_idx".
+
+We use bit indexing from all of the 32 LUTs to store each bit of the
+word, i.e., the first UDS word consists of bit 0 from each 32 LUTs,
+the second word consists of bit 1 from each of the 32 LUTs, and so
+on... for a maximum storage of 16 words.
+
```
Comment (secworks) [inline hw/application_fpga/core/uds/README.md:41]: See, synch with README for the UDI. This is slightly different from what is proposed for the UDI.

File: hw/application_fpga/core/uds/README.md [hunk 4]
```diff
@@ -31,4 +31,15 @@ Any access to another address will be ignored by the core.
...
+The most significant bit in the LUT address serves as the read-enable
+input. This requires the lower half of initialization bits to be
+forced to zero, ensuring that the memory outputs zero when the
+read-enable signal is inactive.
```
Comment (secworks) [inline hw/application_fpga/core/uds/README.md:45]: Possibly not talk about read enable. We want to be able to read UDS words once per power cycle. Since the lower 8 bits in each LUT4 contains the UDS words, the upper 8 bits are set to zero. By setting bit four in the LUT4 address we force reading of bits in the LUT4 to this area, thus getting all zero data returned to the CPU. Or something similar.

We are in fact reading and the read enable (RE) signal is a bad naming. It will be changed. But with the suggested rephrasing such a change will not affect the description.

File: hw/application_fpga/tools/patch_uds_udi.py [hunk 5]
```diff
@@ -5,32 +5,37 @@
...
+# with this environment:
 #
+# - UDS_HEX: path to the UDS file, typically the path to
+#   ../data/uds.hex
+# - UDI_HEX: path to the UDI file, typically the path to ../data/udi.hex
+# - OUT_ASC: path to the ASC output that is then used by icebram and icepack.
 #
-#=======================================================================
+# The script changes the UDS and UDI that are stored in named 4-way
+# LUT (LUT4) instances in the JSON file so we can generate device
```
Comment (secworks) [inline hw/application_fpga/tools/patch_uds_udi.py:24]: 4-way could be rephrased to LUT4 instances, each capable of storing 16 bits. way is probably not something everybody understands.

File: hw/application_fpga/tools/patch_uds_udi.py [hunk 6]
```diff
@@ -5,32 +5,37 @@
...
+# unique bitstreams without running the complete flow just to change
+# UDS and UDI. Then we can just run the final bitstream generation
+# from the ASC file.
+#
+# We represent our UDI and UDS values as a number of 32 bit words:
+#
+# - UDI: 2 words.
+# - UDS: 8 words.
+#
+# We reserve 32 named 4-way LUTs *each* to store the data: UDS in
```
Comment (secworks) [inline hw/application_fpga/tools/patch_uds_udi.py:34]: 4-way.

Comment (secworks) [reply hw/application_fpga/tools/patch_uds_udi.py:34]: The MUX is a 16-1 MUX. The address has four bits, i.e. 2^4 == 16. But more accurately is 16 way. Or not use way as expression.

**Review by secworks — APPROVED** [review]
